File: WindowsApplication1/Utilities/Downloader.vb
﻿Imports System.Globalization
Imports System.IO
Imports System.Net
Imports System.Text
Imports Newtonsoft.Json
Imports Newtonsoft.Json.Linq


Public Class Downloader

    Private Shared GamesTxtURL = "http://showtimes.ninja/static/games.txt"
    Private Shared ScheduleAPIURL = "http://statsapi.web.nhl.com/api/v1/schedule?startDate={0}&endDate={1}&expand=schedule.teams,schedule.game.content.media.epg"
    Private Shared ApplicationVersionURL = " http://showtimes.ninja/static/version.txt"

    Private Shared ApplicationVersionFileName As String = "version.txt"
    Private Shared GamesTextFileName As String = "games.txt"

    Private Shared LocalFileDirectory = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)

    Private Shared Function DownloadContents(URL As String) As String
        Dim client As New WebClient
        client.Encoding = Encoding.UTF8 'Support french chars. I'm looking at you Montreal
        Return client.DownloadString(URL)
    End Function
    Private Shared Sub DownloadFile(URL As String, fileName As String, Optional checkIfExists As Boolean = False, Optional overwrite As Boolean = True)
        Dim fullPath As String = Path.Combine(LocalFileDirectory, fileName)


        If (checkIfExists = False) OrElse (checkIfExists AndAlso My.Computer.FileSystem.FileExists(fullPath) = False) Then
            Console.WriteLine("Downloading File: " & URL & " to " & fullPath)
            My.Computer.Network.DownloadFile(URL, fullPath, "", "", False, 10000, overwrite)
        Else
            Console.WriteLine("Status: File aready exists at " & fullPath)

        End If
    End Sub

    Private Shared Function ReadFileContents(fileName As String) As String
        Dim returnValue As String = ""
        Dim filePath As String = Path.Combine(LocalFileDirectory, fileName)
        Try
            Using streamReader As IO.StreamReader = New IO.StreamReader(filePath)
                returnValue = streamReader.ReadToEnd()
            End Using
        Catch ex As Exception
            Console.WriteLine("Error: " & ex.Message)
        End Try

        Return returnValue
    End Function

    Public Shared Function DownloadApplicationVersion() As String

        Console.WriteLine("Checking: Application version")

        DownloadFile(ApplicationVersionURL, ApplicationVersionFileName)
        Return ReadFileContents(ApplicationVersionFileName).Trim()

    End Function

    Public Shared Function DownloadAvailableGames() As List(Of String)

        Console.WriteLine("Checking: Available games")

        DownloadFile(GamesTxtURL, GamesTextFileName)
        Return ReadFileContents(GamesTextFileName).Split(New Char() {vbLf}).ToList()

    End Function



    Public Shared Function DownloadJSONSchedule(startDate As DateTime) As JObject

        Console.WriteLine("Checking: Fetching game schedule for " & startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture))

        Dim returnValue As New JObject

        Dim IsTodaysSchedule = (startDate.Date.Ticks = DateHelper.GetCentralTime.Date.Ticks)
        Dim dateTimeString As String = startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)
        Dim fileName As String = dateTimeString & ".json"
        Dim URL As String = String.Format(ScheduleAPIURL, dateTimeString, dateTimeString)

        Dim data As String = ""

        If IsTodaysSchedule Then
            Console.WriteLine("Status: Downloading todays current schedule from " & URL)
            data = DownloadContents(URL)
        Else
            DownloadFile(URL, fileName, True)
            data = ReadFileContents(fileName)
        End If

        'returnValue = JObject.Parse(fileContent) 'Don't use above, so we can manually parse out datetime since there may be some issues there with different regions

        Dim reader As New JsonTextReader(New StringReader(data))
        reader.DateParseHandling = DateParseHandling.None
        returnValue = JObject.Load(reader)

        Return returnValue

    End Function

End Class

File: NHLGames.WPF/Services/Downloader.cs
using System;
using System.Globalization;
using System.IO;
using System.Net;
using System.Reflection;
using System.Runtime.Serialization.Json;
using System.Text;
using NHLGames.WPF.Objects;
using NHLGames.WPF.Properties;
using NHLGames.WPF.Utilities;

namespace NHLGames.WPF.Services
{
    public class Downloader
    {
        private const string AppUrl = "https://showtimes.ninja/";
        private const string ApiUrl = "http://statsapi.web.nhl.com/api/v1/schedule";
        private const string ScheduleApiurl = ApiUrl + "?startDate={0}&endDate={1}&expand=schedule.teams,schedule.linescore,schedule.game.seriesSummary,schedule.game.content.media.epg";
        private const string AppVersionUrl = AppUrl + "static/version.txt";

        private const string AppChangelogUrl = AppUrl + "static/changelog.txt";

        private const char Backslash = '\\';

        private static string _localFileDirectory = GetLocalFileDirectory();
        
        private static string GetLocalFileDirectory()
        {
            if (true)
            {
                const string dir = "NHLGames";
                const string file = "test.txt";

                string exeStartupPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
                string localAppDataPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) + Backslash;
                string tempPath = Path.GetTempPath();

                if (File.Exists(localAppDataPath + file))
                {
                    _localFileDirectory = localAppDataPath;
                }
                else if (File.Exists(tempPath))
                {
                    _localFileDirectory = tempPath;
                }
                else
                {
                    _localFileDirectory = exeStartupPath;
                }

                if (!Directory.Exists(_localFileDirectory + dir + Backslash))
                {
                    Directory.CreateDirectory(_localFileDirectory + dir + Backslash);
                }

                _localFileDirectory = _localFileDirectory + dir + Backslash;
            }

            return _localFileDirectory;

        }

        private static string DownloadContents(string server, string url)
        {
            WebClient client = new WebClient();
            string content = string.Empty;
            client.Encoding = Encoding.UTF8;
            try
            {
                content = client.DownloadString(url).Trim();
            }
            catch (Exception)
            {
                if (!(server == AppUrl))
                {
                    Console.WriteLine(Resources.msgServerSeemsDown, server);
                }
            }
            return content;
        }

        private static bool DownloadJsonFile(string server, string url, string fileName)
        {
            WebClient client = new WebClient();
            string filePath = Path.Combine(_localFileDirectory, fileName);
            client.Encoding = Encoding.UTF8;
            try
            {
                client.DownloadFile(url, filePath);
            }
            catch (Exception)
            {
                Console.WriteLine(Resources.msgServerNoRespondTryingAgain, server);
                return false;
            }
            return true;
        }

        private static string ReadFileContents(string fileName)
        {
            string returnValue = "";
            string filePath = Path.Combine(_localFileDirectory, fileName);
            try
            {
                using (var streamReader = new StreamReader(filePath))
                {
                    returnValue = streamReader.ReadToEnd();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(Resources.errorGeneral, ex.Message);
            }

            return returnValue;
        }

        public static string DownloadApplicationVersion()
        {
            var appVers = DownloadContents(AppUrl, AppVersionUrl);
            if (appVers.Contains("<html>"))
            {
                appVers = string.Empty;
            }
            return appVers;
        }

        public static string DownloadChangelog()
        {
            string appChangelog = DownloadContents(AppUrl, AppChangelogUrl);
            if (appChangelog.Contains("<html>"))
            {
                appChangelog = string.Empty;
            }
            return appChangelog;
        }

        public static RootObject DownloadJsonSchedule(DateTime startDate, bool refreshing = false)
        {
            string dateTimeString = startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
            string fileName = dateTimeString + ".json";
            string url = string.Format(ScheduleApiurl, dateTimeString, dateTimeString);
            string data;
            string filePath = Path.Combine(_localFileDirectory, fileName);
            string gettingTerm = refreshing ? Resources.msgRefreshing : Resources.msgFetching;

            if (startDate.Date >= DateHelper.GetPacificTime())
            {
                Console.WriteLine(Resources.msgGettingSchedule, gettingTerm, startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture));
                data = DownloadContents(ApiUrl, url);
            }
            else
            {
                if (LookOldJsonFiles(fileName) & !refreshing)
                {
                    Console.WriteLine(Resources.msgFetchingSavedSchedule, startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture), filePath);
                    data = ReadFileContents(fileName);
                }
                else
                {
                    if (DownloadJsonFile(ApiUrl, url, fileName))
                    {
                        Console.WriteLine(Resources.msgDownloadingJsonFile, startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture), filePath);
                        data = ReadFileContents(fileName);
                    }
                    else
                    {
                        Console.WriteLine(Resources.msgGettingSchedule, gettingTerm, startDate.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture));
                        data = DownloadContents(ApiUrl, url);
                    }
                }
            }

            if (data.Equals(string.Empty))
                return null;

            return RootObject.FromJson(data);

            //return Deserialize<RootObject>(data);
        }

        public static bool LookOldJsonFiles(string filename)
        {
            if (File.Exists(_localFileDirectory + filename))
            {
                return File.GetLastAccessTime(filename).AddDays(1) <= DateTime.Now;
            }
            return false;
        }

        public static string Serialize<T>(T obj)
        {
            DataContractJsonSerializer serializer = new DataContractJsonSerializer(obj.GetType());
            MemoryStream ms = new MemoryStream();
            serializer.WriteObject(ms, obj);
            string retVal = Encoding.UTF8.GetString(ms.ToArray());
            return retVal;
        }

        public static T Deserialize<T>(string json)
        {
            T obj = Activator.CreateInstance<T>();
            MemoryStream ms = new MemoryStream(Encoding.Unicode.GetBytes(json));
            DataContractJsonSerializer serializer = new DataContractJsonSerializer(obj.GetType());
            obj = (T)serializer.ReadObject(ms);
            ms.Close();
            return obj;
        }

    }
}
